azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageVersionDaoImpl.java

```diff
@@ -251,6 +258,54 @@ public int createImageVersion(final ImageVersion imageVersion) {
     return imageVersions;
   }
 
+  @Override
+  public List<ImageVersion> getLatestNonActiveVersionByImageTypes(final Set<String> imageTypes)
+      throws ImageMgmtException {
+    List<ImageVersion> imageVersions = new ArrayList<>();
+    try {
+      // Add outer select clause
+      final StringBuilder inClauseBuilder = new StringBuilder();
+      for (int i = 0; i < imageTypes.size(); i++) {
+        inClauseBuilder.append("?,");
+      }
+      inClauseBuilder.setLength(inClauseBuilder.length() - 1);
+      final Map<String, String> valueMap = new HashMap<>();
+      valueMap.put("image_types", inClauseBuilder.toString());
+
+      final Set<String> nonActiveStates = new TreeSet<>();
+      nonActiveStates.add(State.NEW.getStateValue());
+      nonActiveStates.add(State.UNSTABLE.getStateValue());
+      nonActiveStates.add(State.DEPRECATED.getStateValue());
+      final StringBuilder statesInClauseBuilder = new StringBuilder();
+      for (int i = 0; i < nonActiveStates.size(); i++) {
+        statesInClauseBuilder.append("?,");
+      }
+      statesInClauseBuilder.setLength(statesInClauseBuilder.length() - 1);
+      valueMap.put("version_states", statesInClauseBuilder.toString());
+
+      final StrSubstitutor strSubstitutor = new StrSubstitutor(valueMap);
+      final String query = strSubstitutor.replace(SELECT_LATEST_IMAGE_VERSION_QUERY);
+      log.info("Image version getActiveVersionByImageTypes query : " + query);
+      final List<Object> params = new ArrayList<>();
+      final Set<String> imageTypesInLowerCase =
+          imageTypes.stream().map(String::toLowerCase).collect(Collectors.toSet());
+      log.info("imageTypesInLowerCase: " + imageTypesInLowerCase);
+
+      params.addAll(nonActiveStates);
+      params.addAll(imageTypesInLowerCase);
```

Comment:
Why these image types are added two time? Shouldn't there be join for status?

Reply:
Yes image types are added two times because there are two inner queries containing image types in clause.

Reply:
Will the join with status column help for two inner joins?

Reply:
Status and image types both are part of the queries (both inner queries) and they need to be populated in in order to substitute the query parameters. Btw, this is an existing query and is reused for non active states here as well. For more clarity have added one liner comment regarding addition of these parameters.

Reply:
That sounds good. Thank you for answering.